Document `ProjectBuilder` behavior changes in upgrade guide (#36826)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_9.adoc

=== Potential breaking changes

==== `ProjectBuilder` now handles build-scoped locations more consistently

> link:{javadocPath}/org/gradle/testfixtures/ProjectBuilder.html[`ProjectBuilder`] lets you configure a project directory, which `project.projectDir` and `project.rootDir` correctly reflect.

Reviewer: ```suggestion
link:{javadocPath}/org/gradle/testfixtures/ProjectBuilder.html[`ProjectBuilder`] allows you to configure a specific project directory for tests.
```

> However, `project.layout.settingsDirectory` was not derived from that directory and could point outside of `rootDir`, causing file resolution to inadvertently escape the project directory.

Reviewer: ```suggestion
While `project.projectDir` and `project.rootDir` have always respected this setting, `project.layout.settingsDirectory` previously did not.
This discrepancy could cause file resolution to inadvertently escape the project directory.
```

Gradle now anchors the settings search to the configured project directory, so `layout.settingsDirectory`, `projectDir`, and `rootDir` all point to the same location.
Projects created with `ProjectBuilder` are no longer affected by the surrounding build environment.

If your tests relied on `layout.settingsDirectory` pointing outside the configured project directory, you will need to adjust them.

> Even if you did not use `settingsDirectory` directly, you may still observe changes in file resolution.

Reviewer: ```suggestion
Even if you do not use `settingsDirectory` directly, you may still observe changes in file resolution.
```
